sound/soc/sof/intel/hda-dsp.c

```diff
@@ -352,39 +351,10 @@ static int hda_resume(struct snd_sof_dev *sdev)
 		return ret;
 	}
 
-	hda_dsp_ctrl_misc_clock_gating(sdev, false);
-
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
 	/* Reset stream-to-link mapping */
 	list_for_each_entry(hlink, &bus->hlink_list, list)
 		bus->io_ops->reg_writel(0, hlink->ml_addr + AZX_REG_ML_LOSIDV);
```

Comment:
thats seems wrong. You are changing the programming sequence here by resetting things without taking care of the clock gating. Why?

Reply:
Within @lbetlej 's email:
"CGCTL.MISCBDCGE needs to be disabled during controller reset "
So hda_dsp_ctrl_misc_clock_gating are disabled before controller reset and enabled after that.
with regard to "Reset stream-to-link mapping", there is no need to disable CGCTL.MISCBDCGE before that and then enable after.
So there are 2 solution:
1). move "Reset stream-to-link mapping" into hda_dsp_ctrl_init_chip and before enable CGCTL.MISCBDCGE
2). don't need CGCTL.MISCBDCGE operations before and after "Reset stream-to-link mapping". 
With testing 2) on APL/WHL with HDA mode, it seems there is no difference if don't do that.
@lbetlej do we need disable and enable CGCTL.MISCBDCGE for "Reset stream-to-link mapping"?which solution is right in your opinion?

Reply:
@zhuyingjiang you are playing with fire here. The sequence has been used for several years, and it'll take architectural pointers and a lot more experimental evidence to convince me to change something which is not broken.

Reply:
> @zhuyingjiang you are playing with fire here. The sequence has been used for several years, and it'll take architectural pointers and a lot more experimental evidence to convince me to change something which is not broken.

@plbossart let me explain more here, the first attempt to change here is because the code difference with SST, 
In SST the code is:
static int skl_init_chip(struct hdac_bus *bus, bool full_reset)
{
	struct hdac_ext_link *hlink;
	int ret;
	**_skl_enable_miscbdcge_**(bus->dev, _**false**_);
	ret = snd_hdac_bus_init_chip(bus, full_reset);
	/* Reset stream-to-link mapping */
	list_for_each_entry(hlink, &bus->hlink_list, list)
		bus->io_ops->reg_writel(0, hlink->ml_addr + AZX_REG_ML_LOSIDV);
	**_skl_enable_miscbdcge_**(bus->dev, **_true_**);
	return ret;
}

But in our code it is:

#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
	/* reset and start hda controller */
	ret = hda_dsp_ctrl_init_chip(sdev, true);
==>**_hda_dsp_ctrl_misc_clock_gating_**(sdev, **_false_**);
==>hda_dsp_ctrl_link_reset
==>...
==>**_hda_dsp_ctrl_misc_clock_gating_**(sdev, **_true_**);
	**_hda_dsp_ctrl_misc_clock_gating_**(sdev, **_false_**);
	/* Reset stream-to-link mapping */
	list_for_each_entry(hlink, &bus->hlink_list, list)
		bus->io_ops->reg_writel(0, hlink->ml_addr + AZX_REG_ML_LOSIDV);
	**_hda_dsp_ctrl_misc_clock_gating_**(sdev, **_true_**);
	/* enable ppcap interrupt */
	snd_hdac_ext_bus_ppcap_enable(bus, true);
	snd_hdac_ext_bus_ppcap_int_enable(bus, true);
#else

So in our code there will be disable and enable clock gating twice. 

And a more important difference is in probe,
The SST skl_probe_work which called skl_init_chip still do the below steps:
`/* Reset stream-to-link mapping */`
`list_for_each_entry(hlink, &bus->hlink_list, list)`
`	bus->io_ops->reg_writel(0, hlink->ml_addr + AZX_REG_ML_LOSIDV);`

but in our code, there is no such step in probe.

So in conclusion, I think the solution 1) I mentioned above may be better (move "Reset stream-to-link mapping" into **hda_dsp_ctrl_init_chip** and before enable CGCTL.MISCBDCGE inside. the programming sequence goes the same steps with SST), 
and for 2) , it changed a little much and may leads to uncertain results.

I will change the PR according to solution 1) and request for comments.